### DISCUSSION TITLE
What is your Kubernetes setup?

### DISCUSSION THREAD

**Discussion by shykes** [dagger/dagger#257] [General]
You don’t have to use kubernetes with dagger, but they work great together.

There are many ways to setup a kubernetes cluster, and many ways to deploy applications to it. What is your favorite setup and how would you like Dagger to integrate with it?

**Comment by fkautz** [comment]
A few deployments we should consider:

# Local:
## I personally use:
* KinD (Kubernetes in Docker)
* K3d (K3s in Docker, similar to kind but lower resources)
## Other major distributions which we should support include
* Docker Desktop's Kubernetes (bundled in windows and osx)
* Minikube (vm based)
* Microk8s (snap based)

# Cloud
## I have integration tests for Network Service Mesh (networkservicemesh.io) actively running on:
* GKE (google)
* EKS (aws)
* AKS (azure)
* Equinox Metal (bare metal, provisioned with kubeadm)

Other distributions which I have seen in production include:
* hosted openshift
* IBM cloud
* Alibaba
* platform9
* digital ocean (often individual devs who are looking to learn and control cost)

# On Premise
## Deployments I have personally seen or have been notified of by others
* OpenShift
* Rancher
* K3s for edge deployments
* kubeadm (vanilla kubernetes)
* FlatCar (CoreOS successor)

There are a large number of Kubernetes distributions out there. You can see a list of certified Kubernetes distress here: https://www.cncf.io/certification/software-conformance/

Not being listed here doesn't mean it's bad or not worth targeting. I also am forgetful. :)

I highly recommend focusing on kind, gke, eke, and aks in the first phase. Expand out to On-Premise OpenShift (has a different security model from the others, focused on SELinux), then branch out based on community requests.

The key to success is to focus on *vanilla kubernetes.* If it works on kind and across GKE, EKS, and EKS simultaneously, it will work in most environments.
